File: internal/tracker/stats.go
package tracker

import (
	"github.com/meteormin/wuwa-tracker/internal/types"
)

type StatsCalulator struct {
	standardFiveStarResources types.StandardFiveStarResources
}

func NewStatsCalculator(standardFiveStarResources types.StandardFiveStarResources) *StatsCalulator {
	return &StatsCalulator{
		standardFiveStarResources: standardFiveStarResources,
	}
}

// CalculateStats 는 배열의 Record 들을 바탕으로 통계 지표를 계산합니다.
// 입력 배열은 API에서 가져온 기본 순서(최신순)라고 가정합니다.
func (sc *StatsCalulator) CalculateStats(records []types.Record, gachaType types.GachaType) types.Stats {
	stats := types.Stats{
		GachaType:     gachaType.ID,
		GachaName:     gachaType.Name,
		TotalPulls:    len(records),
		BaseRate:      gachaType.BaseRate,
		ExpectedPulls: gachaType.ExpectedPulls,
		FiveStars:     []types.FiveStarRecord{},
		Records:       []types.Record{},
	}

	pity5 := 0
	pity4 := 0

	// API 응답은 최신순(내림차순)이므로 과거부터 스택을 쌓기 위해 뒤집어서 순회합니다.
	for i := len(records) - 1; i >= 0; i-- {
		rec := records[i]

		// 가챠 획득 스택 쌓기
		pity5++
		pity4++

		// 가챠 기록 저장
		stats.Records = append(stats.Records, rec)

		switch rec.QualityLevel {
		case 5:
			isPickUp := true
			// 한정 캐릭터 배너(1)인 경우에만 픽뚫을 판별합니다.
			if gachaType.HasOffBannerDrop {
				isPickUp = !sc.standardFiveStarResources.Contains(rec.ResourceID)
			}
			stats.FiveStars = append(stats.FiveStars, types.FiveStarRecord{
				Name:     rec.Name,
				Time:     rec.Time,
				Pity:     pity5,
				IsPickUp: isPickUp,
			})
			pity5 = 0
		case 4:
			pity4 = 0
		}
	}

	stats.CurrentPity5 = pity5
	stats.CurrentPity4 = pity4

	// 운 점수(Luck Score) 및 관련 통계 지표를 Go 백엔드 측에서 직접 계산합니다.
	fiveStarCount := len(stats.FiveStars)
	if fiveStarCount > 0 {
		stats.HasFiveStar = true
		sumPity := 0
		for _, fs := range stats.FiveStars {
			sumPity += fs.Pity
		}
		stats.AvgPulls = float64(sumPity) / float64(fiveStarCount)
		if stats.TotalPulls > 0 {
			stats.ActualRate = (float64(fiveStarCount) / float64(stats.TotalPulls)) * 100.0
		}
		if stats.AvgPulls > 0 {
			// 운 점수 계산: 픽업 캐릭터 획득 주기(PickUp Cycle) 기반으로 산출
			// 픽업 캐릭터를 획득하기까지 소모된 총 누적 풀(Pulls) 수와 단일 5성 기대치(ExpectedPulls)를 비교
			expectedTotal := 0
			actualTotal := 0
			currentCyclePulls := 0

			for _, fs := range stats.FiveStars {
				currentCyclePulls += fs.Pity
				if !gachaType.HasOffBannerDrop {
					// 상시 또는 무기 가챠처럼 픽뚫이 없는 경우, 모든 5성 획득이 개별 주기 완성
					expectedTotal += gachaType.ExpectedPulls
					actualTotal += currentCyclePulls
					currentCyclePulls = 0
				} else {
					if fs.IsPickUp {
						// 한정 공명자 배너에서 픽업 캐릭터를 뽑은 경우 주기 완성
						expectedTotal += gachaType.ExpectedPulls
						actualTotal += currentCyclePulls
						currentCyclePulls = 0
					}
				}
			}

			// 아직 픽업을 뽑지 못하고 상시 5성(픽뚫) 상태에서 끝난 진행 중인 주기 반영
			if currentCyclePulls > 0 {
				expectedTotal += gachaType.ExpectedPulls
				actualTotal += currentCyclePulls
			}

			if actualTotal > 0 {
				stats.LuckScore = (float64(expectedTotal) / float64(actualTotal)) * 100
			} else {
				stats.LuckScore = 0.0
			}
		}
	} else {
		stats.HasFiveStar = false
		stats.AvgPulls = 0
		stats.ActualRate = 0
		stats.LuckScore = 0
	}

	// 최신 획득 내역이 배열의 앞쪽에 오도록 FiveStars 를 뒤집습니다.
	reverseFiveStars(stats.FiveStars)

	return stats
}

func reverseFiveStars(s []types.FiveStarRecord) {
	for i, j := 0, len(s)-1; i < j; i, j = i+1, j-1 {
		s[i], s[j] = s[j], s[i]
	}
}

File: crates/wuwa-tracker-core/src/stats.rs
use crate::config::Config;
use wuwa_tracker_types::{FiveStarRecord, GachaType, Record, Stats};

#[derive(Debug, Clone)]
pub struct StatsCalculator {
    standard_five_star_resources: Vec<i32>,
    astrite_per_pull: usize,
}

impl StatsCalculator {
    pub fn new(config: &Config) -> Self {
        Self {
            standard_five_star_resources: config.standard_five_star_resources.clone(),
            astrite_per_pull: config.astrite_per_pull,
        }
    }

    pub fn calc(&self, records: &[Record], gacha_type: &GachaType) -> Stats {
        let mut stats = Stats {
            gacha_type: gacha_type.id,
            gacha_name: gacha_type.name.clone(),
            total_pulls: records.len(),
            total_astrite: records.len() * self.astrite_per_pull,
            current_pity5: 0,
            current_pity4: 0,
            base_rate: gacha_type.base_rate,
            expected_pulls: gacha_type.expected_pulls,
            five_stars: Vec::new(),
            records: Vec::new(),
            avg_pulls: 0.0,
            actual_rate: 0.0,
            luck_score: 0.0,
            has_five_star: false,
        };

        let mut pity5 = 0;
        let mut pity4 = 0;

        for record in records.iter().rev() {
            pity5 += 1;
            pity4 += 1;
            stats.records.push(record.clone());

            match record.quality_level {
                5 => {
                    let is_pick_up = !gacha_type.has_off_banner_drop
                        || !self
                            .standard_five_star_resources
                            .contains(&record.resource_id);
                    stats.five_stars.push(FiveStarRecord {
                        name: record.name.clone(),
                        time: record.time.clone(),
                        pity: pity5,
                        is_pick_up,
                    });
                    pity5 = 0;
                }
                4 => pity4 = 0,
                _ => {}
            }
        }

        stats.current_pity5 = pity5;
        stats.current_pity4 = pity4;

        let five_star_count = stats.five_stars.len();
        if five_star_count > 0 {
            stats.has_five_star = true;
            let sum_pity: i32 = stats.five_stars.iter().map(|item| item.pity).sum();
            stats.avg_pulls = sum_pity as f64 / five_star_count as f64;
            if stats.total_pulls > 0 {
                stats.actual_rate = five_star_count as f64 / stats.total_pulls as f64 * 100.0;
            }

            let mut expected_total = 0;
            let mut actual_total = 0;
            let mut current_cycle_pulls = 0;

            for five_star in &stats.five_stars {
                current_cycle_pulls += five_star.pity;
                if !gacha_type.has_off_banner_drop || five_star.is_pick_up {
                    expected_total += gacha_type.expected_pulls;
                    actual_total += current_cycle_pulls;
                    current_cycle_pulls = 0;
                }
            }

            if current_cycle_pulls > 0 {
                expected_total += gacha_type.expected_pulls;
                actual_total += current_cycle_pulls;
            }

            if actual_total > 0 {
                stats.luck_score = expected_total as f64 / actual_total as f64 * 100.0;
            }
        }

        stats.five_stars.reverse();
        stats.records.reverse();
        stats
    }
}
